presto-main/src/test/java/io/prestosql/sql/planner/iterative/rule/TestPushProjectionIntoTableScan.java

```diff
@@ -0,0 +1,247 @@
+            ConnectorTableHandle expectedTableHandle = createTableHandleForMockConnector(TEST_SCHEMA, "projected_" + TEST_TABLE).getConnectorHandle();
+
+            // Create catalog with applyProjection enabled
+            MockConnectorFactory factory = createMockFactory(ImmutableMap.of(inputColumnName, columnHandle), true);
+            ruleTester.createCatalog(MOCK_CATALOG, factory, ImmutableMap.of());
+
+            Metadata metadata = ruleTester.getMetadata();
+            TypeAnalyzer typeAnalyzer = new TypeAnalyzer(new SqlParser(), metadata);
+            PushProjectionIntoTableScan optimizer = new PushProjectionIntoTableScan(metadata, typeAnalyzer);
+
+            // Prepare project node symbols and types
+            Symbol identitySymbol = new Symbol("symbol_identity");
+            Symbol dereferenceSymbol = new Symbol("symbol_dereference");
+            Symbol constantSymbol = new Symbol("symbol_constant");
+            ImmutableMap<Symbol, Type> types = ImmutableMap.<Symbol, Type>builder()
+                    .put(baseColumnSymbol, ROW_TYPE)
+                    .put(identitySymbol, ROW_TYPE)
+                    .put(dereferenceSymbol, BIGINT)
+                    .put(constantSymbol, BIGINT)
+                    .build();
+
+            // Prepare project node assignments
+            ImmutableMap<Symbol, Expression> inputProjections = ImmutableMap.<Symbol, Expression>builder()
+                    .put(identitySymbol, baseColumnSymbol.toSymbolReference())
+                    .put(dereferenceSymbol, new DereferenceExpression(baseColumnSymbol.toSymbolReference(), new Identifier("a")))
+                    .put(constantSymbol, new LongLiteral("5"))
+                    .build();
+            ImmutableMap<Symbol, ConnectorExpression> inputConnectorProjections = inputProjections.entrySet().stream()
+                    .collect(toImmutableMap(e -> e.getKey(), e -> translate(MOCK_SESSION, e.getValue(), typeAnalyzer, viewOf(types)).get()));
+
+            // Expected output projections and column handles after applyProjection has been invoked
+            ImmutableMap<Symbol, String> connectorProjectionNames = inputConnectorProjections.entrySet().stream()
+                    .collect(toImmutableMap(entry -> entry.getKey(), entry -> entry.getValue().toString()));
+            ImmutableMap<Symbol, Expression> expectedOutputProjections = ImmutableMap.<Symbol, Expression>builder()
+                    .put(identitySymbol, new SymbolReference("projected_variable_" + connectorProjectionNames.get(identitySymbol)))
+                    .put(dereferenceSymbol, new SymbolReference("projected_dereference_" + connectorProjectionNames.get(dereferenceSymbol)))
+                    .put(constantSymbol, new SymbolReference("projected_constant_" + connectorProjectionNames.get(constantSymbol)))
+                    .build();
+            ImmutableMap<String, ColumnHandle> expectedColumnHandles = ImmutableMap.<String, ColumnHandle>builder()
```

Comment:
As I said earlier, it would be easier to pass assertion function, not the actual fully fledged object here.